Feature support for handling terms of service for users

divpatel10 opened the pull request:
This supports platform or subsites terms of service

divpatel10 committed:
feat: Feature support for handling terms of service for users

This supports platform or subsites terms of service

divpatel10 committed:
fix(test-case-failure): Fix pylint failures and added test.py file

Fixed pylint failures and added test.py file for TermsOfService Feature

ztraboo commented on openedx/features/termsofservice/static/termsofservice/jsx/Modal.jsx:
```diff
@@ -0,0 +1,124 @@
+            </div>
+
+            <div className="modal-footer d-inline text-center">
+              <form>
```
@divpatel10 
Please look into adding the `{% csrf_token %}` with the top of the `<form>` element. See more about how this is done here. https://docs.djangoproject.com/en/3.2/ref/csrf/#how-to-use-it

Maybe this is overboard since there really isn't form information being passed to the platform but it would be an additional security measure for CSRF.

ztraboo replied:
Maybe we don't need to do what I set above seeing that the check is handled in this class. Please advise.

I think we probably want to add this `csrf_token` should the form ever call an API endpoint like this https://github.com/CUCWD/edx-platform/pull/201/files#diff-21f5be93e9938068f8a5c39358cb195f64d77f94239fc729b0edadfdb1471235R13-R14. This API is done though in this class' constructor and not done through a form by the user where there are input fields.

I think this `@csrf_exempt` was the correct call for your API endpoint that you called in `retrievePage()`.
https://docs.djangoproject.com/en/3.2/ref/csrf/#django.views.decorators.csrf.csrf_exempt

divpatel10 replied:
I added the csrf_token when a POST request is made to the API

ztraboo replied:
@divpatel10 
I'm not sure that you were wrong in having ​@csrf_exempt​ decorator.
[Comparing 67795329991e5e57764e8f88a36990a179259412..924951973b97fe6fa06b43114fbcc031cbb33ffc · CUCWD/edx-platform (github.com)](https://github.com/CUCWD/edx-platform/compare/67795329991e5e57764e8f88a36990a179259412..924951973b97fe6fa06b43114fbcc031cbb33ffc)



Looking at the documentation here [Cross Site Request Forgery protection | Django documentation | Django (djangoproject.com)](https://docs.djangoproject.com/en/3.2/ref/csrf/#django.views.decorators.csrf.csrf_exempt) it appears that if you add that decorator to the view method that it is exempt from protection from the CSRF middleware.

So I guess removing this decorator has the same effect? Not sure honestly.

My thought is that you would add the `{% csrf_token %}` to the form at the top like so, 
```
<form method="post">{% csrf_token %}
```
then add the `@requires_csrf_token` decorator to this method.
```
from django.views.decorators.csrf import ensure_csrf_cookie

@ensure_csrf_cookie
def terms_of_service_api(request): # lint-amnesty, pylint: disable=missing-function-docstring
```

divpatel10 committed:
feat: Code Review Changes

In regards to CSRF Token

divpatel10 committed:
feat: Code Review Changes in regards to CSRF Token

divpatel10 committed:
feat: Code Review changes

To improve error handling

divpatel10 committed:
feat: Code Review changes for more error handling

ztraboo committed:
feat: Added `tos_exists_for_site` to api to not show modal when TermsOfService is empty for site.

ztraboo committed:
feat(terms-of-service): Added `tos_exists_for_site` to TOS api 

Added `tos_exists_for_site` to API to not show the Modal when TOS All Sites is empty

ztraboo commented:
- Update logic in `edx-platform/openedx/features/termsofservice/api/v1/views.py:terms_of_service_api` to be all sites or subsite specific with `if settings.FEATURES.get('ENABLE_TERMSOFSERVICE_PER_SUBSITE')` statement.
- Figure out spacing in Terms of Service between bullet points.
- Figure out why the LMS Dashboard page reloads after the user selects.
- Create migration that inserts initial TermsOfService.

divpatel10 committed:
feat: Code Review changes for Terms of Service

Added Django migration to add initial data, Fixed CSS issues with TOS, Fixed double redirect after user agrees to TOS, and refactored some code for readability

divpatel10 committed:
feat: Code Review changes for Terms of Service

Added Django migration to add initial data, Fixed CSS issues with TOS, Fixed double redirect after user agrees to TOS, and refactored some code for readability

divpatel10 commented:
Curf ID: cm-2023-01

TOS HTML Text:
```<h1>Terms of Service</h1>
<p>
    The Terms and Conditions (hereinafter, the "Terms") contained herein shall govern all usage of the EducateWorkforce<sup>TM</sup> portal located at the EducateWorkforce<sup>TM</sup> website (www.EducateWorkforce.com), including all pages of the EducateWorkforce website (hereinafter, collectively, the "Website"). These Terms apply in full force and effect to your use of the Website.  By clicking "Accept", below, and by using the Website, you expressly accept all terms and conditions contained herein in full. You must not use the Website, if you have any objection to any of these Terms.
</p>

<h3 class="tos-part">Part 1. Usage and Access of Content.</h3>
<ul class="tos-section">
    <li class="tos-section-item"><span>1.1</span><p>By using the Website and accepting these Terms, you expressly acknowledge the rights of Clemson University Research Foundation ("CURF") and Clemson University ("Clemson") in the workforce educational program entitled EducateWorkforce<sup>TM</sup>, in all materials and content on the Website, or otherwise accessible through use of the Website, including all intellectual property rights to the foregoing, whether under copyright, trademark, trade secret, etc., (collectively, the "Content"), and agrees to do nothing inconsistent with such rights or to challenge such rights.</p></li>
    <li class="tos-section-item"><span>1.2</span><p>You further agree to use the Content solely for non-commercial educational purposes, and only in the form and manner provided through the Website and with appropriate copyright notices and trademark legends as prescribed on the Website, or otherwise by CURF, and not to use any other trademark or service mark of CURF or Clemson in combination with any Content without prior written approval of CURF.</p></li>
    <li class="tos-section-item"><span>1.3</span><p>You further agree that any usage of any trademarks, service marks, or tradenames within the Content (collectively, "Marks") by Licensee shall inure to the benefit of and be on behalf of CURF and Clemson.</p></li>
    <li class="tos-section-item"><span>1.4</span><p>You further agree that the nature and quality of all services rendered by you in connection with Content (including the Marks), and any related advertising, promotional and other related uses of Content (including the Marks) by you, shall be in support of your non-commercial educational use of the Content.</p></li>
    <li class="tos-section-item"><span>1.5</span><p>You further agree that through use of the Website, no right, title or interest in the Content, other than the right to use the Content in accordance with these Terms, is provided.</p></li>
    <li class="tos-section-item"><span>1.6</span><p>You further agree that title to any improvements, modifications, corrections, or derivative works of the Content shall vest solely in Clemson.</p></li>
    <li class="tos-section-item"><span>1.7</span><p>You further agree to cooperate with CURF in facilitating CURF’s control of the nature and quality of any Content or services provided incorporating the Marks, and to supply CURF with information of all uses of the Content upon request by CURF or Clemson, as well as all other information reasonably requested by CURF or Clemson for its assessment, evaluation and/or research regarding the outcomes and effectiveness of EducateWorkforce<sup>TM</sup>. You acknowledge and agree that all information and data provided to CURF and/or Clemson in accordance with these Terms may be provided to and used by Clemson in connection with EducateWorkforce<sup>TM</sup> and its other educational, scientific, and charitable activities.</p></li>
    <li class="tos-section-item"><span>1.8</span><p>You further agree to comply with all applicable laws and regulations and to obtain all appropriate government approv¬als pertaining to the use of the Content.</p></li>
    <li class="tos-section-item"><span>1.9</span><p>You further agree to only access the Content through the website EducateWorkforce.com.</p></li>
    <li class="tos-section-item"><span>1.10</span><p>You further agree that CURF may restrict or cease your access to the Website in response to any violation of any provision within the Terms.</p></li>
</ul>

<h3 class="tos-part">Part 2. Use Restrictions.</h3>
<ul class="tos-section">
    <li class="tos-section-item"><span>2.1</span><p>You are expressly and emphatically restricted from all of the following:</p>
        <ul class="tos-section-alpha">
            <li><p>publishing any Content in any media, except as expressly authorized by the Website;</p></li>
            <li><p>selling, sublicensing and/or otherwise commercializing any Content;</p></li>
            <li><p>using the Website in any way that is, or may be, damaging to the Website;</p></li>
            <li><p>using the Website in any way that impacts user access to the Website;</p></li>
            <li><p>using the Website contrary to applicable laws and regulations, or in a way that causes, or may cause, harm to the Website, or to any person or business entity;</p></li>
            <li><p>engaging in any data mining, data harvesting, data extracting or any other similar activity in relation to the Website, or while using the Website; and</p></li>
            <li><p>using the Website to engage in any advertising or marketing not expressly permitted by the Website.</p></li>
        </ul>
    </li>
    <li class="tos-section-item"><span>2.2</span><p>Certain areas of the Website may be restricted from access by you and CURF or Clemson may further restrict access by you to any areas of the Website, at any time, in its sole and absolute discretion.  Any user ID and password you may have for the Website are confidential and you must maintain confidentiality of such information.</p></li>
    <li class="tos-section-item"><span>2.3</span><p>The Website is not for use by users under the age of 13 years old.  You must not use the Website if you are under the age of 13 years old.  Neither CURF nor Clemson knowingly collects any personal information from children under the age of 13.  By using this website, you agree that you are at least 18 years old or that you possess consent from your parent, legal guardian, or education institution to use this website.</p></li>
</ul>

<h3 class="tos-part">Part 3. No warranties.</h3>
<p>
    THE WEBSITE IS PROVIDED "AS IS," WITH ALL FAULTS, AND NEITHER CURF NOR CLEMSON MAKE ANY EXPRESS OR IMPLIED REPRESENTATIONS OR WARRANTIES, OF ANY KIND RELATED TO THE WEBSITE OR THE MATERIALS CONTAINED ON THE WEBSITE, INCLUDING ANY WARRANTIES OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Additionally, nothing contained on the Website shall be construed as providing consult or advice to you.
</p>

<h3 class="tos-part">Part 4. Limitation of liability.</h3>
<p>
    In no event shall CURF or Clemson, or any of its officers, directors and employees, be liable to you for anything arising out of or in any way connected with your use of the Website, whether such liability is under contract, tort or otherwise, and neither CURF nor Clemson, nor any of its officers, directors and employees shall not be liable for any indirect, consequential or special liability arising out of or in any way related to your use of the Website.
</p>

<h3 class="tos-part">Part 5. Indemnification.</h3>
<p>
    You hereby agree to indemnify to the fullest extent CURF and Clemson from and against any and all liabilities, costs, demands, causes of action, damages and expenses (including reasonable attorney’s fees) arising out of or in any way related to your breach of any of the provisions of these Terms.
</p>

<h3 class="tos-part">Part 6. Severability.</h3>
<p>
    If any provision of these Terms is found to be unenforceable or invalid under any applicable law, such unenforceability or invalidity shall not render these Terms unenforceable or invalid as a whole, and such provisions shall be deleted without affecting the remaining provisions herein.
</p>

<h3 class="tos-part">Part 7. Variation of Terms.</h3>
<p>
    CURF is permitted to revise these Terms at any time as it sees fit, and by using the Website you are expected to review such Terms on a regular basis to ensure you understand and agree to all terms and conditions governing use of the Website.
</p>

<h3 class="tos-part">Part 8. Assignment.</h3>
<p>
    CURF shall be permitted to assign, transfer, and subcontract its rights and/or obligations under these Terms without any notification or consent required. However, you shall not be permitted to assign, transfer, or subcontract any of your rights and/or obligations under these Terms.
</p>

<h3 class="tos-part">Part 9. Entire Agreement.</h3>
<p>
    These Terms, including any legal notices and disclaimers contained on the Website, constitute the entire agreement between CURF and you in relation to your use of the Website and the Content, with the exception of any EducateWorkforce<sup>TM</sup> License Agreement.  With the exception of the terms and agreement in any executed EducateWorkforce<sup>TM</sup> License Agreement, these Terms supersede all prior agreements and understandings with respect to the same.
</p>

<h3 class="tos-part">Part 10. Governing Law &amp; Jurisdiction.</h3>
<p>
    These Terms will be governed by and construed in accordance with the laws of the State of South Carolina, and you submit to the non-exclusive jurisdiction of the state and federal courts located in South Carolina for the resolution of any disputes.
</p>```

ztraboo committed:
feat: Allow `TermsOfServiceAllSites` to edit existing records.

ztraboo committed:
feat: Make sure that `TermsOfService` modal form doesn't refresh the page.

ztraboo committed:
feat: Add some error handling when `TermsOfService` models are empty.
